src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -103,6 +111,251 @@ void CodeGen::genSSEIntrinsic(GenTreeHWIntrinsic* node)
+            genConsumeRegs(op1);
+
+            argList = argList->Rest();
+            op2     = argList->Current();
+            op2Reg  = op2->gtRegNum;
+            genConsumeRegs(op2);
+
+            argList = argList->Rest();
+            op3     = argList->Current();
+            op3Reg  = op3->gtRegNum;
+            genConsumeRegs(op3);
+
+            argList = argList->Rest();
+            op4     = argList->Current();
+            op4Reg  = op4->gtRegNum;
+            genConsumeRegs(op4);
+
+            regNumber tmpReg = node->GetSingleTempReg();
+
+            emit->emitIns_SIMD_R_R_R(INS_unpcklps, tmpReg, op2Reg, op1Reg, TYP_SIMD16);
+            emit->emitIns_SIMD_R_R_R(INS_unpcklps, targetReg, op4Reg, op3Reg, TYP_SIMD16);
+            emit->emitIns_SIMD_R_R_R_I(INS_shufps, targetReg, targetReg, tmpReg, 68, TYP_SIMD16);
+            break;
+        }
+
+        case NI_SSE_SetAll:
+            assert(baseType == TYP_FLOAT);
+            assert(op2 == nullptr);
+            emit->emitIns_SIMD_R_R_R_I(INS_shufps, targetReg, op1Reg, op1Reg, 0, TYP_SIMD16);
+            break;
+
+        case NI_SSE_SetZero:
+            assert(baseType == TYP_FLOAT);
+            assert(op1 == nullptr);
+            assert(op2 == nullptr);
+            emit->emitIns_SIMD_R_R_R(INS_xorps, targetReg, targetReg, targetReg, TYP_SIMD16);
+            break;
+
+        case NI_SSE_Shuffle:
+        {
+            GenTreeArgList* argList;
+            ssize_t         ival;
+
+            // Shuffle takes 3 operands, so op1 should be an arg list with two
+            // additional node in the chain.
+            assert(baseType == TYP_FLOAT);
+            assert(op1->OperIsList());
+            assert(op1->AsArgList()->Rest() != nullptr);
+            assert(op1->AsArgList()->Rest()->Rest() != nullptr);
+            assert(op1->AsArgList()->Rest()->Rest()->Rest() == nullptr);
+            assert(op2 == nullptr);
+
+            argList = op1->AsArgList();
+            op1     = argList->Current();
+            op1Reg  = op1->gtRegNum;
+            genConsumeRegs(op1);
+
+            argList = argList->Rest();
+            op2     = argList->Current();
+            op2Reg  = op2->gtRegNum;
+            genConsumeRegs(op2);
+
+            argList = argList->Rest();
+            op3     = argList->Current();
+            ival    = op3->AsIntConCommon()->IconValue();
+            genConsumeRegs(op3);
+
+            emit->emitIns_SIMD_R_R_R_I(INS_shufps, targetReg, op1Reg, op2Reg, (int)ival, TYP_SIMD16);
+            break;
+        }
+
+        case NI_SSE_Sqrt:
+            assert(baseType == TYP_FLOAT);
+            assert(op2 == nullptr);
+            emit->emitIns_SIMD_R_R(INS_sqrtps, targetReg, op1Reg, TYP_SIMD16);
+            break;
+
+        case NI_SSE_StaticCast:
+            assert(op2 == nullptr);
+            if (op1Reg != targetReg)
+            {
+                emit->emitIns_SIMD_R_R(INS_movaps, targetReg, op1Reg, TYP_SIMD16);
```

Comment:
I think `NI_SSE_StaticCast` should not reach codgen that should be converted to `CopyBLK` in the importer. Actually, even `copy` may be unnecessary because `Vector128/256<T>` is immutable types.

Reply:
> Actually, even copy may be unnecessary because Vector128/256<T> is immutable types.

Yes, it looks like it's not necessary. It's all `TYP_SIMD16` even if the base type is `float` or `int`.

Reply:
This is exactly how `GT_BITCAST` works. The codegen is needed for when the `StaticCast` results in assignment rather than immediate use.

For the case of `var y = MyOp(StaticCast(x))`, the JIT will set `op1Reg` and `targetReg` to the same, and consume the value directly (no code will be generated).

However, for the case of `var y = StaticCast(x)`, the JIT has to store the value to the local (assuming it isn't itself optimized away) and must generate code to do the store.

Reply:
> This is exactly how GT_BITCAST works

AFAIK bitcast is intended to allow you to treat a type as another type (e.g. treat `float` as `int`). That is not what `StaticCast` is doing.

Reply:
See where I set `isTgtPref` in the lsra here: https://github.com/dotnet/coreclr/pull/15538/commits/c75bdb73ef833e96ad6b3b7cc300362fe4518af6#diff-498fd4859cb147c9bf082f5ebf32ca8fR2554

Reply:
I know that but that is not the point. The point is that `StaticCast` should be a no-op at IR level, it's purely a language level construct that allows you to reinterpret `Vector128<float>` as `Vector128<int>`. Due to the way these types are represented in IR it is not the same thing as reinterpreting `float` as `int`.

Reply:
> AFAIK bitcast is intended to allow you to treat a type as another type (e.g. treat float as int). That is not what StaticCast is doing.

How is that not what `StaticCast` is doing? `StaticCast` is taking a `Vector128<T>` and treating it as a `Vector128<U>`. In both the case of GT_BITCAST and StaticCast, we are taking one set of bits and reinterpreting it.

> should be a no-op

Yes, unless you are storing the value.
```
var x = Sse.SetScalar(4.0f);
var y = Sse.StaticCast<float, int>(x);
x = Sse.SqrtScalar(x);
```

In the above code, we are creating another local and then modifying the first. Therefore, `StaticCast` is not a no-op, it is a store. `y` needs to remain the integer representation of `4.0f` and `x` needs to remain `2.0f`, so that any subsequent code executes correctly.

Reply:
Ex: The following program
```C++
#include <intrin.h>
#include <iostream>

int main()
{
	auto x = _mm_set_ss(4.0f);
	std::cout << x.m128_f32[0] << ", " << x.m128_f32[1] << ", " << x.m128_f32[2] << ", " << x.m128_f32[3] << "\r\n";

	auto y = _mm_castps_si128(x);
	std::cout << y.m128i_i32[0] << ", " << y.m128i_i32[1] << ", " << y.m128i_i32[2] << ", " << y.m128i_i32[3] << "\r\n";

	x = _mm_sqrt_ss(x);
	std::cout << x.m128_f32[0] << ", " << x.m128_f32[1] << ", " << x.m128_f32[2] << ", " << x.m128_f32[3] << "\r\n";
	std::cout << y.m128i_i32[0] << ", " << y.m128i_i32[1] << ", " << y.m128i_i32[2] << ", " << y.m128i_i32[3] << "\r\n";

    return 0;
}
```

prints:
```
4, 0, 0, 0
1082130432, 0, 0, 0
2, 0, 0, 0
1082130432, 0, 0, 0
```

Reply:
> In both the case of GT_BITCAST and StaticCast, we are taking one set of bits and reinterpreting it.

And reinterpreting basically means doing nothing, the bit pattern does not change. At best, you're copying the bit pattern from a float register to an integer register. And reinterpretation makes sense only when talking about different JIT types - TYP_FLOAT and TYP_INT for example. Otherwise there's nothing to reinterpret.

In the case of SSE the type is the same (it's either TYP_SIMD16 or TYP_SIMD32) and the registers are the same. Keep in mind that `Vector128<float>` and `Vector128<int>` both map to TYP_SIMD16 and Y/XMM registers.

> Yes, unless you are storing the value.

`StaticCast` doesn't store values anywhere. It produces a value that a separate node (a GT_LCL_VAR, GT_IND or another intrinsic node) may store somewhere. Neither `GT_BITCAST` stores anything anywhere.

Reply:
I'm still not convinced here. SIMD is also doing roughly the same (albeit without setting `isTgtPref` in `lsra`) for the `Vector<T>.op_explicit` calls, which are the equivalent to the `StaticCast` calls here.

I also don't see any existing examples in the importer (although I may have missed them), where the type of reinterpreting you are wanting is being done. Would you be able to point one out?

Reply:
> I'm still not convinced here. SIMD is also doing roughly the same (albeit without setting isTgtPref in lsra)

Leave LSRA alone, focus on the IR. You have a node that produces a TYP_SIMD16 value. You have another node that uses the TYP_SIMD16 value. Why should there be a GT_BITCAST node between those 2 nodes?

> or the Vector<T>.op_explicit calls, which are the equivalent to the StaticCast calls here.

For all I know that code could be useless. It's not the first time when it happens to discover that some piece of JIT code is somehow useless, broken, inefficient etc.

> I also don't see any existing examples in the importer (although I may have missed them), where the type of reinterpreting you are wanting is being done. Would you be able to point one out?

Yeah, I suppose that doesn't exist, it's an unusual case really. I would expect that doing a pop and push of the popped value should work fine.

If it doesn't work then see if you can use a NOP, it's closer to the meaning of `StaticCast` than BITCAST. Besides, I don't think BITCAST has ever been used in HIR so you may encounter problems if you try to use it in the importer.

Reply:
Just pushing/popping the stack doesn't work, as it detects an unbalanced stack.

Using `gtNewNothingNode` (NOP) doesn't work because the subsequent STLOC instruction expects a GenTree that has the appropriate return type (and gtOp1 needs to carry the value that should be stored).

Returning a NOP node with the appropriate return type fails multiple asserts down the line (such as in `impAssignStructPtr` as it isn't supported and requires more extensive changes.

I am going to keep what I have right now, for now. It is small, simple, produces the correct result, produces (what appears to be) good codegen, and results in a NOP in the places we expect it to. Additionally, similar examples to it can also be found elsewhere in the codebase, so it isn't something new.

Reply:
OK, I'll take a closer look after this is merged. I'm too lazy to get your branch and experiment with it :grin:

Reply:
To be clear, I'm perfectly happy changing this to something better, it just doesn't appear to be readily supported right now (and I don't want to waste a bunch of time trying to figure it out -- at least not when I was able to follow an existing example and get it working already) 😄

Reply:
> it just doesn't appear to be readily supported right now

Yeah, it could be, especially GT_NOPs with struct types. It's not clear why pop/push doesn't work but that's something that can be investigated in the future.

Reply:
Push/Pop doesn't work because of the handling in `impImportCall` here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L8199 (callRetType is grabbed from the method sig here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L7047).

We don't have a VOID return type, so we try to push the call node onto the stack here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L8199

`impPushOnStack` itself validates that we don't go over `compMaxStack` (https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L83), which in this case is 1.